accepted/benchmark-infrastructure.md

```diff
@@ -0,0 +1,102 @@
+show historical and current performance data.
+
+# Motivation
+[motivation]: #motivation
+
+As we develop the Cranelift compiler and its machine backends, and as it is
+used in more production settings, its performance becomes increasingly
+important. In many settings, a few percentage points change in compilation time
+or generated-code performance is quite a big deal. In addition, experience in
+many compiler projects over the years has shown that small performance deltas
+have a way of accumulating over time; if there is no systematic tracking, this
+gradual drift can add up to sizeable performance regressions.
+
+It's thus important both to track historical trends, in order to avoid
+regressions and (more positively) to measure our progress; and to measure the
+effect of individual changes, when significant enough, to determine objectively
+whether optimizations or design choices are good ideas.
+
+Independently of *what* benchmarks we measure, we need an *infrastructure* to
+measure, record, and visualize statistics. This RFC proposes building such an
+infrastructure.
+
+# Proposal sketch
+[proposal]: #proposal
+
+## Step 1: Establish physical infrastructure
+
+First, we need to establish a set of machines to run benchmarks and host
+results. The GitHub-hosted CI infrastructure we currently use for testing will
+likely not be adequate, because (i) runtime of benchmarks will be longer than
+we would like for CI actions, (ii) we don't want to tie benchmark runs to CI
+runs, necessarily, and (iii) CI-runner environments are unpredictable and
+unreliable environments for accurate, low-noise benchmark measurements.
+
+We already have access to one AArch64 server-grade host for interactive
+development of the Cranelift ARM backend, generously [provided by
+ARM](https://github.com/WorksOnArm/cluster/issues/204). We may be able to use
+this host for running benchmarks as well, if all stakeholders agree.
+
+We will need an x86-64 host as well; the best option is likely to rent a
+low-cost dedicated host. Details of hosting choice and funding are TBD.
+
+Relevant Bytecode Alliance-affiliated engineers would then be granted access to
+these machines and maintain a benchmark-runner infrastructure as described
+below.
+
+## Step 2: Build and host a metrics-visualization web UI
+
+We will need a frontend to present the benchmarking data that this
+infrastructure collects. Johnnie Birch and Andrew Brown of Intel have developed
+a prototype web-based frontend that records and plots performance of benchmark
+suites over time; this is the most likely candidate. As long as traffic is low
+enough, we can plan to use the benchmark-runner host(s) as web hosts as well in
+order to serve this UI. Alternately, if hosting performance becomes an issue,
+we could design a static/pregenerated-page workflow wherein the runners upload
+data to a git repository and static data is served from GitHub Pages (or
+another static host).
```

Comment:
For a dynamic website one of my concerns would be management with deployment and such, but one possibility is that I believe Heroku has a free tier which should be more than sufficient for this purpose?

Reply:
That's a good question; @jlb6740 or @abrown might have more thoughts on what is possible here with the current Sightglass UI?

Reply:
I think we can get pretty far with a static github pages frontend and have JS to switch between static graph images that were pre-rendered by the benchmark analysis.

Reply:
@cfallin @alexcrichton .. Not sure I get the question correctly (specifically static data?). There is a viewer (nuxt.js) that loads a json file that servers as the database. Separately there is another html file that does a simple plot of the latest data (can be extended to allow selecting any history of data, any specific data). What was attempted before was collecting data on one machine and the scp this updated json database history file to github pages that served the same viewer. I don't remember the issues with that but I can try to duplicate that effort again and go from there.